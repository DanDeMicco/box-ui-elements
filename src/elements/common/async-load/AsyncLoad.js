// @flow
import * as React from 'react';
import retry from 'utils/retry';

type Props = {
    loader: () => Promise<any>,
    fallback?: React.Element<any>,
    errorComponent?: React.ComponentType<any>,
};

type State = {
    error: ?Error,
};

const NullComponent = () => null;

const AsyncLoad = (asyncProps: Props) => {
    const LazyComponent = React.lazy(() => retry(() => asyncProps.loader()));

    return class extends React.Component<Object, State> {
        state = {
            error: null,
        };

        static getDerivedStateFromError(error: Error) {
            return {
                error,
            };
        }

        render() {
            const { fallback = null, errorComponent: ErrorComponent = NullComponent } = asyncProps;
            const { error } = this.state;
            if (error) {
                return <ErrorComponent error={error} />;
            }

            return (
                <React.Suspense fallback={fallback}>
                    <LazyComponent {...this.props} />
                </React.Suspense>
            );
        }
    };
};

export default AsyncLoad;
