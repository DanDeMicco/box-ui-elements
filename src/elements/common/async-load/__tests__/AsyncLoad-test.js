import * as React from 'react';
import { shallow } from 'enzyme';
import AsyncLoad from '../AsyncLoad';

describe('elements/common/async-load/AsyncLoad', () => {
    const defaultProps = {
        loader: jest.fn(),
    };

    const getAsyncComponent = (props = defaultProps) => AsyncLoad(props);

    test('should return a react component', () => {
        const AsyncComponent = getAsyncComponent();
        expect(AsyncComponent.prototype).toBeInstanceOf(React.Component);
    });

    test('should load the lazy component', () => {
        const AsyncComponent = getAsyncComponent();
        const wrapper = shallow(<AsyncComponent foo="bar" />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render the error component if there is an error', () => {
        const errorComponent = () => <div>ERROR!!</div>;
        const AsyncComponent = getAsyncComponent({
            ...defaultProps,
            errorComponent,
        });

        const wrapper = shallow(<AsyncComponent />);
        wrapper.setState({
            error: new Error('foo'),
        });

        expect(wrapper).toMatchSnapshot();
    });
});
