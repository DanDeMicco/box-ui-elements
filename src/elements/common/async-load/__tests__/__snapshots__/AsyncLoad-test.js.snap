// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`elements/common/async-load/AsyncLoad should load the lazy component 1`] = `
<React.Suspense
  fallback={null}
>
  <React.Lazy
    foo="bar"
  />
</React.Suspense>
`;

exports[`elements/common/async-load/AsyncLoad should render the error component if there is an error 1`] = `
<errorComponent
  error={[Error: foo]}
/>
`;
